capactio/capact
# Handle input TypeInstances

**Description**

Changes proposed in this pull request:

- Added injecting Input TypeInstances when provided

**Testing**

Execute following scenario:

* Create and run action to deploy Postgresql
* Create `cap.type.database.postgresql.config` TypeInstance for this Postgresql
* Create and run action to deploy Jira. Pass it created TypeInstance
* Check that Jira is using passed Postgresql instance. 

Mutations and queries to create the Postgresql Action:

```graphql
mutation CreateStandaloneDB {
    createAction(
        in: {
            name: "standalone-postgress"
            actionRef: {
                path: "cap.interface.database.postgresql.install"
                revision: "0.1.0"
            }
            dryRun: false
            advancedRendering: false
            input: {
              parameters: "{\"superuser\":{\"username\":\"postgres\",\"password\":\"secret\"},\"defaultDBName\": \"test\"}"             
            }
        }
    ) {
        name
          input {
          parameters
        }
    }
}

query GetStandaloneDBAction {
  action(name:"standalone-postgress") {
    name
    createdAt
    input {
      parameters
    }
    status{
      phase
    }
  }
}

mutation RunStandaloneDBAction{
    runAction(name: "standalone-postgress") {
        name
    }
}

mutation DeleteStandaloneDBAction{
  deleteAction(name:"standalone-postgress") {
    name
    createdAt
    input {
      parameters
    }
  }
}
```

Queries and Mutation to manage TypeInstance:

```graphql
mutation CreatePostgresqlTypeInstance {
    createTypeInstance(
      in:{
        alias:"postgresql",
        typeRef:{
          path: "cap.type.database.postgresql.config"
          revision:"0.1.0"
        }
        value: {defaultDBName: "test", host: "postgresql-1613738969", port: "5432",superuser:{password: "secret",username: "postgres"}}
      }
        
    ) {
    		spec{
          typeRef {
            path
            revision
          }
          value
        }
    }
}

query GetTypeInstances {
  typeInstances{
    metadata{
      id
    }
    spec{
      typeRef{
        path
        revision
      }
      value
    }
  }
}
```

After creating TypeInstace, list it and copy ID.
Replace TypeInstance ID in the following query and deploy Jira.

```graphql
mutation CreateJiraWithPreinstalledPostgress {
    createAction(
        in: {
            name: "jira"
            actionRef: {
                path: "cap.interface.productivity.jira.install"
                revision: "0.1.0"
            }
            dryRun: false
            advancedRendering: false
            input: {
              parameters: "{\"host\":\"jira\"}"
              typeInstances:{
                name: "postgresql",
                id:"16158168-bdad-4185-9bf5-a3338d1820d0"
              }
            }
        }
    ) {
        name
    		input {
          parameters
          typeInstances{
            name
            id
          }
        }
    }
}

query GetJiraWithPreinstalledPostgress {
  action(name:"jira") {
    name
    createdAt
    input {
      parameters
    }
    status{
      phase
      runner{
        status
      }
    }
  }
}

mutation RunJiraWithPreinstalledPostgress{
    runAction(name: "jira") {
        name
    }
}

mutation DeleteJiraWithPreinstalledPostgress{
  deleteAction(name:"jira") {
    name
    createdAt
    input {
      parameters
    }
  }
}
```


Run `helm list` to see that there is only one instance of PostgreSQL and one of Jira
